library/Faktory/JobState.hs

```diff
@@ -0,0 +1,53 @@
+module Faktory.JobState
+  ( JobState(..)
+  ) where
+
+import Faktory.Prelude
+
+import Control.Arrow ((&&&))
+import Control.Error.Util (note)
+import Data.Aeson
+import Data.HashMap.Strict (HashMap)
+import qualified Data.HashMap.Strict as HashMap
+import qualified Data.Text as T
+
+data JobState
+  = JobStateUnknown
+  | JobStateEnqueued
+  | JobStateWorking
+  | JobStateSuccess
+  | JobStateFailed
+  | JobStateDead
+  deriving stock (Eq, Show, Bounded, Enum)
+
+instance ToJSON JobState where
+  toJSON = toJSON . jobStateToText
+  toEncoding = toEncoding . jobStateToText
+
+instance FromJSON JobState where
+  parseJSON = withText "JobState" $ either fail pure . jobStateFromText
+
+jobStateToText :: JobState -> Text
+jobStateToText = \case
+  JobStateUnknown -> "unknown"
+  JobStateEnqueued -> "enqueued"
+  JobStateWorking -> "working"
+  JobStateSuccess -> "success"
+  JobStateFailed -> "failed"
+  JobStateDead -> "dead"
+
+jobStateFromText :: Text -> Either String JobState
+jobStateFromText x =
+  note
+      (unpack
+      $ "Invalid JobState: "
+      <> x
+      <> ", must be one of "
+      <> T.intercalate ", " (HashMap.keys jobStateMap)
+      )
+    $ HashMap.lookup x jobStateMap
+
+jobStateMap :: HashMap Text JobState
+jobStateMap =
+  HashMap.fromList $ map (jobStateToText &&& id) [minBound .. maxBound]
+{-# NOINLINE jobStateMap #-}
```

Comment:
This is one of the many things I'm a little weak at is inline pragmas. Could you elucidate why we want this?

Reply:
I just do what Chris P tells me.

My loose understanding is that this function is effectively a constant, and GHC will redefine it as such and never actually need to runtime-evaluate anything -- as long as its definition doesn't get in-lined to call sites. GHC may be smart enough to do that anyway, but this pragma ensures it.

A super unnecessary micro-optimization IMO, but also an easy and innocuous one.